### PR TITLE
Get measurements

### LINKED ISSUE #81: Backend: Get Measurements
- Switch to TimescaleDB
- Create a table with timescale features for the measuremenets
- Include filter methods (use timescale native functions)
- "As a user i want to see my own smart meter data"

### PR DESCRIPTION
**Pull request by 1r-1na**
## Check hypertables
- with command `select * from _timescaledb_catalog.hypertable` in a tenant db it is possiblee to check for hypertables.
image

### CODE REVIEW

File: src/SMAIAXBackend.API/Endpoints/Measurement/GetMeasurementsEndpoint.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+namespace SMAIAXBackend.API.Endpoints.Measurement;
+
+public static class GetMeasurementsEndpoint
+{
+    public static async Task<Ok<List<MeasurementRawDto>>> Handle(IMeasurementListService measurementListService,
+        [FromQuery] Guid smartMeterId,
+        [FromQuery] DateTime? startAt, [FromQuery] DateTime? endAt)
+    {
+        var measurements =
+            await measurementListService.GetFilteredMeasurementsByTenantAndSmartMeterAsync(
```
Comment (Smightym8) [inline src/SMAIAXBackend.API/Endpoints/Measurement/GetMeasurementsEndpoint.cs]: Rename `GetFilteredMeasurementsByTenantAndSmartMeterAsync` to `GetFilteredMeasurementsBySmartMeterAsync`.

File: src/SMAIAXBackend.API/Middlewares/ExceptionHandlerMiddleware.cs [hunk 2]
```diff
@@ -42,6 +42,7 @@ public async ValueTask<bool> TryHandleAsync(
                 problemDetails.Status = StatusCodes.Status409Conflict;
                 problemDetails.Title = "Conflict";
                 break;
+            case ArgumentException:
```
Comment (Smightym8) [inline src/SMAIAXBackend.API/Middlewares/ExceptionHandlerMiddleware.cs]: Remove the `ArgumentException` and use only custom exceptions here.

File: src/SMAIAXBackend.Application/Services/Implementations/MeasurementListService.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+using SMAIAXBackend.Domain.Model.ValueObjects.Ids;
+using SMAIAXBackend.Domain.Repositories;
+
+namespace SMAIAXBackend.Application.Services.Implementations;
+
+public class MeasurementListService(
+    IMeasurementRepository measurementRepository,
+    ISmartMeterRepository smartMeterRepository) : IMeasurementListService
+{
+    public async Task<List<MeasurementRawDto>> GetFilteredMeasurementsByTenantAndSmartMeterAsync(
```
Comment (Smightym8) [inline src/SMAIAXBackend.Application/Services/Implementations/MeasurementListService.cs]: Rename `GetFilteredMeasurementsByTenantAndSmartMeterAsync` to `GetFilteredMeasurementsBySmartMeterAsync`.

File: src/SMAIAXBackend.Application/Services/Interfaces/IMeasurementListService.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    ///     Get all or filtered measurements of a tenant by smart meter.
+    /// </summary>
+    /// <param name="smartMeterId">The specific smart meter id.</param>
+    /// <param name="startAt">
+    ///     Optional timestamp filter. Data with a timestamp newer/greater than or equal to startAt will be
+    ///     returned.
+    /// </param>
+    /// <param name="endAt">Optional timestamp filter. Data with a timestamp older/smaller than endAt will be returned.</param>
+    /// <returns>All measurements between given limitations or by default measurements of one day.</returns>
+    Task<List<MeasurementRawDto>> GetFilteredMeasurementsByTenantAndSmartMeterAsync(Guid smartMeterId,
```
Comment (Smightym8) [inline src/SMAIAXBackend.Application/Services/Interfaces/IMeasurementListService.cs]: Rename `GetFilteredMeasurementsByTenantAndSmartMeterAsync` to `GetFilteredMeasurementsBySmartMeterAsync`.

File: src/SMAIAXBackend.Domain/Repositories/IMeasurementRepository.cs [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    ///     Reads all measurements of a tenant or filtered measurements when the parameter is given.
+    /// </summary>
+    /// <param name="smartMeterId">The specific smart meter id.</param>
+    /// <param name="startAt">
+    ///     Optional timestamp filter. Data with a timestamp newer/greater than or equal to startAt will be
+    ///     returned.
+    /// </param>
+    /// <param name="endAt">Optional timestamp filter. Data with a timestamp older/smaller than endAt will be returned.</param>
+    /// <returns>All measurements between given limitations or by default measurements of one day.</returns>
+    Task<List<Measurement>> GetMeasurementsByTenantAndSmartMeterAsync(SmartMeterId smartMeterId,
```
Comment (Smightym8) [inline src/SMAIAXBackend.Domain/Repositories/IMeasurementRepository.cs]: Rename `GetMeasurementsByTenantAndSmartMeterAsync` to `GetMeasurementsBySmartMeterAsync`.

File: src/SMAIAXBackend.Infrastructure/Repositories/MeasurementRepository.cs [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+public class MeasurementRepository(
+    TenantDbContext tenantDbContext,
+    ILogger<MeasurementRepository> logger) : IMeasurementRepository
+{
+    public async Task<List<Measurement>> GetMeasurementsByTenantAndSmartMeterAsync(SmartMeterId smartMeterId,
+        DateTime? startAt = null,
+        DateTime? endAt = null)
+    {
+        if (endAt > startAt)
```
Comment (Smightym8) [inline src/SMAIAXBackend.Infrastructure/Repositories/MeasurementRepository.cs]: Move this check to the application service and throw a custom exception.

File: src/SMAIAXBackend.Infrastructure/Repositories/MeasurementRepository.cs [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+        {
+            logger.LogError("Argument 'endAt' must be before 'startAt'.");
+            throw new ArgumentException("Argument 'endAt' must be before 'startAt'.");
+        }
+
+        // If endAt is null, by default one month is returned, either from startAt or from DateTime.UtcNow.
+        endAt ??= startAt?.AddDays(-1) ?? DateTime.UtcNow.AddDays(-1);
```
Comment (Smightym8) [inline src/SMAIAXBackend.Infrastructure/Repositories/MeasurementRepository.cs]: Move this to the application service.

**Review by Smightym8 — CHANGES_REQUESTED** [review]

File: src/SMAIAXBackend.Infrastructure/Repositories/MeasurementRepository.cs (hunk 7 above)
Comment (1r-1na) [reply src/SMAIAXBackend.Infrastructure/Repositories/MeasurementRepository.cs]: I disabled the automatic range to one day and made startAt and endAt required.

**Review by Smightym8 — APPROVED** [review]

**Pull request merged**